src/backend/cdb/cdbmutate.c

```diff
@@ -2610,12 +2619,114 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
 	return true;
 }
 
+static bool
+root_slice_is_executed_on_coordinator(Plan *plan, PlannerInfo *root)
+{
+	Query	   *query = root->parse;
+
+	if (query->parentStmtType != PARENTSTMTTYPE_NONE)
+	{
+		/*
+		 * For CTAS, SELECT INTO, COPY INTO, REFRESH MATERIALIZED VIEW, the
+		 * root slice is not executed on the coordinator.
+		 */
+		return false;
+	}
+
+	if (IsA(plan, ModifyTable))
+	{
+		ModifyTable *mt = (ModifyTable *) plan;
+
+		if (list_length(mt->resultRelations) > 0)
+		{
+			ListCell   *lc = list_head(mt->resultRelations);
+			Oid			reloid = getrelid(lfirst_int(lc), query->rtable);
+
+			if (GpPolicyFetch(reloid)->ptype != POLICYTYPE_ENTRY)
+			{
+				/*
+				 * For the modification node by Postgres planner, if the table
+				 * distribution policy is not master-only, then the root slice
+				 * is not executed on the coordinator.
+				 */
+				return false;
+			}
+		}
+	}
+
+	if (IsA(plan, DML))
+	{
+		DML		   *dml = (DML *) plan;
+		Oid			reloid = getrelid(dml->scanrelid, query->rtable);
+
+		if (GpPolicyFetch(reloid)->ptype != POLICYTYPE_ENTRY)
+		{
+			/*
+			 * For the modification node by ORCA planner, if the table
+			 * distribution policy is not master-only, then the root slice is
+			 * not executed on the coordinator.
+			 */
+			return false;
+		}
+	}
+
+	if (plan->flow)
```

Comment:
[Here](https://github.com/arenadata/gpdb/pull/498/files#diff-a4950bc4a9e9552824f7ba4f6bf300b64335fb668074e51f7b9ee3ac30fe7a78R2671-R2673) I removed the `else`, because the executor always checks for parallel cursors at the end.